# 응답과 예외 매핑

앞에서 우리는 인터셉터를 통해 응답과 발생한 예외를 잡아 변형을 가할 수 있다고 했다. 각각에 대한 간단한 인터셉터를 예를 들어 살펴보겠다. 먼저 전달받은 응답에 변형 해보자. 라우터 핸들러에서 전달한 응답을 객체로 감싸서 전달하도록 하는 TransformInterceptor를 만들었다.

```ts
import { CallHandler, ExecutionContext, Injectable, NestInterceptor } from '@nestjs/common';
import { Observable } from 'rxjs';
import { map } from 'rxjs/operators';

export interface Response<T> {
  data: T;
}

@Injectable()
export class TransformInterceptor<T> implements NestInterceptor<T, Response<T>> {
  intercept(context: ExecutionContext, next: CallHandler): Observable<Response<T>> {
    return next
      .handle()
      .pipe(map(data => {
        return { data }
      }));
  }
}
```

TransformInterceptor는 LoggingInterceptor와 다르게 Generic으로 타입 T를 선언하고 있다. NestInterceptor 인터페이스의 정의를 보면 Generic으로 T, R 타입 2개를 선언하도록 되어 있다. 사실 둘다 기본이 any 타입이기 때문에 어떤 타입이 와도 상관없다. T는 응답 스트림을 지원하는 Observable 타입이어야 하고, R은 응답의 값을 Observable로 감싼 타입을 정해 주어야 한다. 우리는 타입스크립트를 사용하고 있으므로 타입을 명확히 지정해 주면 더 안전하게 코딩을 할 수 있다. TransformInterceptor의 예로 다시 돌아가서 T는 any타입이 될 것이고, R은 Response를 지정했다. Response는 우리의 요구사항에 맞게 정의한 타입, 즉 data 프로퍼티를 가지는 객체가 되도록 강제한다.

이제 TransformInterceptor를 전역으로 적용해 보자. useGlobalInterceptors에 콤마로 인터셉터 객체를 추가하면 된다.

```ts
...
import { TransformInterceptor } from './transform.interceptor';

async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  app.useGlobalInterceptors(
    new LoggingInterceptor(),
    new TransformInterceptor(),
  );
  await app.listen(3000);
}
bootstrap();
```

그러면 요청을 보내서 결과를 살펴볼자!

```bash
$ curl http://localhost:3000/users
{
    "data":"This action returns all users"
}
```

서버 콘솔에 LoggingInterceptor의 로그도 잘 남는 지 확인해보자.

이제 라우트 핸들링 도중 던져진 예외를 잡아서 변환해 보는 예를 보도록 하자. 발생한 모든 에러를 잡아서 502 Bad Gateway로 변경하는 예를 만들어 보겠다. 이것은 사실 좋은 예가 아니다. 예외를 변환하는 것은 예외 필터에서 다루는 것이 더 낫다는 생각입니다만 인터셉터를 이용해서도 가능하다는 정도로 알아두도록 하자!

```ts
import { Injectable, NestInterceptor, ExecutionContext, BadGatewayException, CallHandler } from '@nestjs/common';
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';

@Injectable()
export class ErrorsInterceptor implements NestInterceptor {
  intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
    return next
      .handle()
      .pipe(
        catchError(err => throwError(() => new BadGatewayException())),
      );
  }
}
```

이번에는 전역으로 적용하지 않고 라우트 핸들러에 GET /users/:id 엔드포인트에만 적용해 보자. 강제로 500 에러를 일으키도록 했다.

```ts
@UseInterceptors(ErrorsInterceptor)
@Get(':id')
findOne(@Param('id') id: string) {
  throw new InternalServerErrorException();
}
```

이제 해당 엔드포인트로 요청을 해 보면 아래 처럼 500 에러가 502로 바뀌어 나가는 것을 볼 수 있다.

```bash
$ curl http://localhost:3000/users/1
{
    "statusCode":502,
    "message":"Bad Gateway"
}
```